File: togo/internal/cli/show.go
package cli

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"slices"
	"strings"

	"github.com/nikeee/cappu/internal/audit"
	"github.com/nikeee/cappu/internal/config"
	"github.com/nikeee/cappu/internal/lockfile"
	"github.com/nikeee/cappu/internal/packages"
	"github.com/nikeee/cappu/internal/sources"
)

// severityStyle maps a severity to its display colour. Port of SEVERITY_STYLE
// in src/cli/show.ts (critical is bold+red, rendered as nested paints).
var severityStyle = map[audit.Severity]string{
	audit.SeverityCritical: "red",
	audit.SeverityHigh:     "red",
	audit.SeverityModerate: "yellow",
	audit.SeverityLow:      "cyan",
	audit.SeverityUnknown:  "dim",
}

const labelWidth = 13

// searchHint is appended to the "bad coordinate" / "not found" errors: both
// mean the user does not have an exact coordinate yet, and `cappu search` is
// how to find one. Port of SEARCH_HINT in src/cli/show.ts.
const searchHint = "search for a package with `cappu search <query>`"

var whitespaceRe = regexp.MustCompile(`\s+`)

// projectShow is how group:artifact relates to the current project.
type projectShow struct {
	configurations []string
	declared       string // "" when not declared
	installed      string // "" when not locked
}

// showData is everything the card (and --json) shows, gathered once. Port of
// ShowData in src/cli/show.ts.
type showData struct {
	groupID, artifactID, version string
	explicitVersion              bool
	latestVersion                string // "" when no versions are published
	versionCount                 int
	newer                        int
	description                  string
	homepage                     string
	scmURL                       string
	spdx                         []string
	rawLicenses                  []string
	dependencies                 []packages.DependencyDeclaration
	project                      projectShow
	vulnerabilities              []audit.Advisory
}

// showError is a message to print and the exit code to use.
type showError struct {
	message string
	code    int
}

// parseCoordinate splits "group:artifact[:version]" into parts, ok=false if malformed.
func parseCoordinate(coord string) (groupID, artifactID, version string, ok bool) {
	parts := strings.Split(coord, ":")
	if len(parts) < 2 || len(parts) > 3 {
		return "", "", "", false
	}
	for _, p := range parts {
		if p == "" {
			return "", "", "", false
		}
	}
	if len(parts) == 3 {
		version = parts[2]
	}
	return parts[0], parts[1], version, true
}

// listVersionsAcross returns the versions from the first source that has any.
func listVersionsAcross(groupID, artifactID string, srcs []packages.PackageSource) []string {
	for _, s := range srcs {
		versions, err := s.ListVersions(groupID, artifactID)
		if err == nil && len(versions) > 0 {
			return versions
		}
	}
	return nil
}

// metadataAcross returns the metadata from the first source that has it.
func metadataAcross(c packages.Coordinates, srcs []packages.PackageSource) *packages.PackageMetadata {
	for _, s := range srcs {
		if meta, err := s.GetMetadata(c); err == nil && meta != nil {
			return meta
		}
	}
	return nil
}

// gatherProjectContext reports where (if anywhere) this project depends on key.
func gatherProjectContext(cfg *config.Config, key string) projectShow {
	byName := map[string]map[string]string{
		"api":                 cfg.Dependencies.API,
		"implementation":      cfg.Dependencies.Implementation,
		"annotationProcessor": cfg.Dependencies.AnnotationProcessor,
		"testImplementation":  cfg.Dependencies.TestImplementation,
	}
	pc := projectShow{configurations: []string{}}
	for _, name := range config.Configurations {
		if v, ok := byName[name][key]; ok {
			pc.configurations = append(pc.configurations, name)
			if pc.declared == "" {
				pc.declared = v
			}
		}
	}
	if lock := lockfile.Read(cfg); lock != nil {
		all := slices.Concat(lock.Packages, lock.ProcessorPackages, lock.TestPackages)
		for _, p := range all {
			c := p.Coords()
			if string(c.Key()) == key {
				pc.installed = string(c.Version)
				break
			}
		}
	}
	return pc
}

// buildShowData gathers the full detail of one package, or a showError to print.
// Port of buildShowData in src/cli/show.ts.
func buildShowData(coord string, cfg *config.Config, srcs []packages.PackageSource, auditSource audit.AuditSource) (*showData, *showError) {
	groupID, artifactID, wantVersion, ok := parseCoordinate(coord)
	if !ok {
		return nil, &showError{
			message: "show needs group:artifact[:version], e.g. `cappu show com.google.code.gson:gson`; " + searchHint,
			code:    2,
		}
	}

	versions := listVersionsAcross(groupID, artifactID, srcs)
	latest := ""
	if len(versions) > 0 {
		latest = versions[len(versions)-1]
	}
	version := wantVersion
	if version == "" {
		version = latest
	}
	if version == "" {
		return nil, &showError{message: fmt.Sprintf("package not found: %s:%s; %s", groupID, artifactID, searchHint), code: 1}
	}
	c := packages.NewCoordinates(groupID, artifactID, version)

	meta := metadataAcross(c, srcs)
	if meta == nil && len(versions) == 0 {
		return nil, &showError{message: fmt.Sprintf("package not found: %s:%s:%s; %s", groupID, artifactID, version, searchHint), code: 1}
	}

	// OSV scan of just this version; a network failure must not sink the card.
	var vulnerabilities []audit.Advisory
	if report, err := audit.AuditPackages([]packages.Coordinates{c}, auditSource); err == nil && len(report.Vulnerable) > 0 {
		vulnerabilities = slices.Clone(report.Vulnerable[0].Advisories)
		// OSV returns advisories in database order; show them worst severity first.
		slices.SortStableFunc(vulnerabilities, func(a, b audit.Advisory) int {
			return slices.Index(audit.SeverityOrder, a.Severity) - slices.Index(audit.SeverityOrder, b.Severity)
		})
	}

	newer := 0
	if idx := slices.Index(versions, version); idx >= 0 {
		newer = len(versions) - 1 - idx
	}

	data := &showData{
		groupID:         groupID,
		artifactID:      artifactID,
		version:         version,
		explicitVersion: wantVersion != "",
		latestVersion:   latest,
		versionCount:    len(versions),
		newer:           newer,
		project:         gatherProjectContext(cfg, groupID+":"+artifactID),
		vulnerabilities: vulnerabilities,
	}
	if meta != nil {
		// POM descriptions are often multi-line and indented; collapse to one line.
		data.description = strings.TrimSpace(whitespaceRe.ReplaceAllString(meta.Description, " "))
		data.homepage = meta.Homepage
		data.scmURL = meta.ScmURL
		seen := map[string]bool{}
		for _, l := range meta.Licenses {
			data.rawLicenses = append(data.rawLicenses, l.Name)
			if id, ok := packages.NormalizeLicense(l.Name, l.URL); ok && !seen[string(id)] {
				seen[string(id)] = true
				data.spdx = append(data.spdx, string(id))
			}
		}
		data.dependencies = slices.Clone(meta.Dependencies)
		slices.SortFunc(data.dependencies, func(a, b packages.DependencyDeclaration) int {
			return strings.Compare(
				string(a.GroupID)+":"+string(a.ArtifactID),
				string(b.GroupID)+":"+string(b.ArtifactID),
			)
		})
	}
	return data, nil
}

// --- JSON output (one stable shape for --json) -------------------------------

type showDepJSON struct {
	GroupID    string `json:"groupId"`
	ArtifactID string `json:"artifactId"`
	Version    string `json:"version"`
	Scope      string `json:"scope,omitempty"`
	Optional   bool   `json:"optional,omitempty"`
}

type showAdvisoryJSON struct {
	ID            string   `json:"id"`
	Aliases       []string `json:"aliases"`
	Severity      string   `json:"severity"`
	Summary       string   `json:"summary"`
	FixedVersions []string `json:"fixedVersions"`
	URL           string   `json:"url"`
}

type showProjectJSON struct {
	Configurations []string `json:"configurations"`
	Declared       *string  `json:"declared"`
	Installed      *string  `json:"installed"`
}

type showJSON struct {
	GroupID         string             `json:"groupId"`
	ArtifactID      string             `json:"artifactId"`
	Version         string             `json:"version"`
	LatestVersion   *string            `json:"latestVersion"`
	VersionCount    int                `json:"versionCount"`
	Description     *string            `json:"description"`
	Homepage        *string            `json:"homepage"`
	ScmURL          *string            `json:"scmUrl"`
	License         []string           `json:"license"`
	Dependencies    []showDepJSON      `json:"dependencies"`
	Project         showProjectJSON    `json:"project"`
	Vulnerabilities []showAdvisoryJSON `json:"vulnerabilities"`
}

// nilIfEmpty returns a *string that is nil (JSON null) for the empty string.
func nilIfEmpty(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

func showToJSON(d *showData) showJSON {
	license := d.spdx
	if len(license) == 0 {
		license = d.rawLicenses
	}
	if license == nil {
		license = []string{}
	}
	deps := make([]showDepJSON, 0, len(d.dependencies))
	for _, dep := range d.dependencies {
		deps = append(deps, showDepJSON{
			GroupID:    string(dep.GroupID),
			ArtifactID: string(dep.ArtifactID),
			Version:    string(dep.Version),
			Scope:      string(dep.Scope),
			Optional:   dep.Optional,
		})
	}
	vulns := make([]showAdvisoryJSON, 0, len(d.vulnerabilities))
	for _, a := range d.vulnerabilities {
		aliases := a.Aliases
		if aliases == nil {
			aliases = []string{}
		}
		fixed := a.FixedVersions
		if fixed == nil {
			fixed = []string{}
		}
		vulns = append(vulns, showAdvisoryJSON{
			ID:            string(a.ID),
			Aliases:       aliases,
			Severity:      string(a.Severity),
			Summary:       a.Summary,
			FixedVersions: fixed,
			URL:           a.URL,
		})
	}
	return showJSON{
		GroupID:       d.groupID,
		ArtifactID:    d.artifactID,
		Version:       d.version,
		LatestVersion: nilIfEmpty(d.latestVersion),
		VersionCount:  d.versionCount,
		Description:   nilIfEmpty(d.description),
		Homepage:      nilIfEmpty(d.homepage),
		ScmURL:        nilIfEmpty(d.scmURL),
		License:       license,
		Dependencies:  deps,
		Project: showProjectJSON{
			Configurations: d.project.configurations,
			Declared:       nilIfEmpty(d.project.declared),
			Installed:      nilIfEmpty(d.project.installed),
		},
		Vulnerabilities: vulns,
	}
}

// --- text card ---------------------------------------------------------------

// renderShowCard renders the coloured detail card (paint is a no-op when colour
// is disabled). Port of renderShowCard in src/cli/show.ts.
func renderShowCard(d *showData, paint func(format, text string) string) string {
	var lines []string
	row := func(label, value string) {
		lines = append(lines, "  "+paint("dim", fmt.Sprintf("%-*s", labelWidth, label))+value)
	}

	// Header: coordinates + version, with a freshness hint from the version list.
	hint := ""
	switch {
	case !d.explicitVersion || d.version == d.latestVersion:
		hint = paint("green", "latest")
	case d.newer > 0:
		hint = paint("yellow", fmt.Sprintf("%d newer available", d.newer))
	}
	header := paint("cyan", paint("bold", d.groupID+":"+d.artifactID)) + " " + paint("bold", d.version)
	if hint != "" {
		header += "  " + hint
	}
	lines = append(lines, header)
	if d.description != "" {
		lines = append(lines, paint("dim", "  "+d.description))
	}
	lines = append(lines, "")

	switch {
	case len(d.spdx) > 0:
		row("License", paint("cyan", strings.Join(d.spdx, ", ")))
	case len(d.rawLicenses) > 0:
		row("License", paint("yellow", strings.Join(d.rawLicenses, ", ")+" (no SPDX id)"))
	default:
		row("License", paint("dim", "none declared"))
	}
	if d.homepage != "" {
		row("Homepage", paint("blue", d.homepage))
	}
	if d.scmURL != "" {
		row("Repository", paint("blue", d.scmURL))
	}
	if d.latestVersion != "" {
		row("Versions", d.latestVersion+" "+paint("dim", "(latest)")+paint("dim", fmt.Sprintf(", %d published", d.versionCount)))
	}
	row("In project", formatProject(d.project, paint))

	lines = append(lines, "", "  "+paint("bold", fmt.Sprintf("Dependencies (%d)", len(d.dependencies))))
	if len(d.dependencies) == 0 {
		lines = append(lines, "    "+paint("dim", "none"))
	} else {
		for _, dep := range d.dependencies {
			lines = append(lines, "    "+formatDependency(dep, paint))
		}
	}

	lines = append(lines, "", "  "+paint("bold", "Vulnerabilities"))
	if len(d.vulnerabilities) == 0 {
		lines = append(lines, "    "+paint("green", "no known vulnerabilities"))
	} else {
		for _, a := range d.vulnerabilities {
			cve := ""
			if len(a.Aliases) > 0 {
				cve = " (" + strings.Join(a.Aliases, ", ") + ")"
			}
			fixed := ""
			if len(a.FixedVersions) > 0 {
				fixed = paint("dim", "  [fixed in: "+strings.Join(a.FixedVersions, ", ")+"]")
			}
			sev := paint(severityStyle[a.Severity], strings.ToUpper(string(a.Severity)))
			if a.Severity == audit.SeverityCritical {
				sev = paint("bold", sev) // critical is bold+red
			}
			lines = append(lines, "    "+sev+"  "+string(a.ID)+cve+" - "+a.Summary+fixed)
			lines = append(lines, "      "+paint("dim", a.URL))
		}
		lines = append(lines, "    "+paint("dim", "run `cappu audit` to scan the whole dependency tree"))
	}

	return strings.Join(lines, "\n") + "\n"
}

func formatProject(pc projectShow, paint func(format, text string) string) string {
	if len(pc.configurations) == 0 {
		return paint("dim", "not a direct dependency")
	}
	where := strings.Join(pc.configurations, ", ")
	var parts []string
	if pc.declared != "" {
		parts = append(parts, "declared "+pc.declared)
	}
	if pc.installed != "" {
		parts = append(parts, "installed "+pc.installed)
	}
	if len(parts) > 0 {
		where += paint("dim", " ("+strings.Join(parts, ", ")+")")
	}
	return where
}

func formatDependency(d packages.DependencyDeclaration, paint func(format, text string) string) string {
	coord := string(d.GroupID) + ":" + string(d.ArtifactID) + ":" + string(d.Version)
	var tags []string
	if d.Scope != "" && d.Scope != "compile" {
		tags = append(tags, string(d.Scope))
	}
	if d.Optional {
		tags = append(tags, "optional")
	}
	if len(tags) > 0 {
		coord += paint("dim", "  "+strings.Join(tags, ", "))
	}
	return coord
}

// RunShow handles `cappu show <coord>`: a single-package detail card. With
// jsonOut, the same data is emitted machine-readable. Port of src/cli/show.ts.
func RunShow(coord string, cfg *config.Config, jsonOut bool) int {
	auditSource := audit.NewOsvSource(audit.CachedFetchJSON(nil))
	data, showErr := buildShowData(coord, cfg, sources.Configured(cfg), auditSource)
	if showErr != nil {
		fmt.Fprintf(os.Stderr, "cappu: %s\n", showErr.message)
		return showErr.code
	}

	if jsonOut {
		b, _ := json.MarshalIndent(showToJSON(data), "", "  ")
		fmt.Fprintf(os.Stdout, "%s\n", b)
	} else {
		fmt.Fprint(os.Stdout, renderShowCard(data, painter(os.Stdout)))
	}
	if len(data.vulnerabilities) > 0 {
		return 1
	}
	return 0
}

File: src/cli/show.ts
// `cappu show <group:artifact[:version]>`: a single-package detail card -
// coordinates, description, license, homepage/repo, published versions, the
// declared dependencies, how this project depends on it, and any known
// vulnerabilities (OSV). The npm/cargo `info`/`show` shape. --json emits the
// same data machine-readable (implied under an agent, like the other commands).

import type { CappuConfig } from "../config.ts";
import {
  type AuditSource,
  auditPackages,
  cachedFetchJson,
  OsvSource,
  type Severity,
  SEVERITY_ORDER,
} from "../audit/index.ts";
import { configuredSources, type ProjectContext, projectContext } from "../install.ts";
import {
  type Coordinates,
  type DependencyDeclaration,
  normalizeLicense,
  type PackageMetadata,
  type PackageSource,
  toCoordinates,
} from "../packages/index.ts";
import { type StyleFormat, painter } from "./style.ts";

const SEVERITY_STYLE: Record<Severity, StyleFormat> = {
  critical: ["bold", "red"],
  high: "red",
  moderate: "yellow",
  low: "cyan",
  unknown: "dim",
};

const LABEL_WIDTH = 13;

interface ShowAdvisory {
  readonly id: string;
  readonly aliases: readonly string[];
  readonly severity: Severity;
  readonly summary: string;
  readonly fixedVersions: readonly string[];
  readonly url: string;
}

/** Everything the card (and --json) shows, gathered once so the two can't drift. */
export interface ShowData {
  readonly groupId: string;
  readonly artifactId: string;
  readonly version: string;
  /** Whether the user pinned the version (vs. defaulting to latest). */
  readonly explicitVersion: boolean;
  readonly latestVersion: string | undefined;
  readonly versionCount: number;
  /** How many published versions are newer than the shown one. */
  readonly newer: number;
  readonly description: string | undefined;
  readonly homepage: string | undefined;
  readonly scmUrl: string | undefined;
  /** SPDX ids the licenses map to, else the raw POM names. */
  readonly spdx: readonly string[];
  readonly rawLicenses: readonly string[];
  readonly dependencies: readonly DependencyDeclaration[];
  readonly project: ProjectContext;
  readonly vulnerabilities: readonly ShowAdvisory[];
}

interface ShowError {
  readonly error: string;
  readonly code: number;
}

/** Split "group:artifact[:version]" into its parts, or undefined if malformed. */
function parseCoordinate(
  coord: string,
): { groupId: string; artifactId: string; version?: string } | undefined {
  const parts = coord.split(":");
  if (parts.length < 2 || parts.length > 3 || parts.some(p => p === "")) return undefined;
  const [groupId, artifactId, version] = parts;
  return { groupId: groupId!, artifactId: artifactId!, version };
}

/** The first source that lists any version of group:artifact wins (oldest first). */
async function listVersionsAcross(
  groupId: string,
  artifactId: string,
  sources: readonly PackageSource[],
): Promise<string[]> {
  for (const source of sources) {
    const versions = await source.listVersions(groupId, artifactId);
    if (versions.length > 0) return versions;
  }
  return [];
}

/** The first source that has metadata for these coordinates wins. */
async function metadataAcross(
  coordinates: Coordinates,
  sources: readonly PackageSource[],
): Promise<PackageMetadata | undefined> {
  for (const source of sources) {
    const metadata = await source.getMetadata(coordinates);
    if (metadata) return metadata;
  }
  return undefined;
}

/** Gather the full detail of one package, or an error to print (and its exit code). */
export async function buildShowData(
  coord: string,
  config: CappuConfig,
  sources: readonly PackageSource[],
  auditSource: AuditSource,
): Promise<ShowData | ShowError> {
  const parsed = parseCoordinate(coord);
  if (!parsed) {
    return {
      error: "show needs group:artifact[:version], e.g. `cappu show com.google.code.gson:gson`",
      code: 2,
    };
  }
  const { groupId, artifactId } = parsed;

  const versions = await listVersionsAcross(groupId, artifactId, sources);
  const latest = versions.at(-1);
  const version = parsed.version ?? latest;
  if (version === undefined) {
    return { error: `package not found: ${groupId}:${artifactId}`, code: 1 };
  }
  const coordinates = toCoordinates(groupId, artifactId, version);

  const metadata = await metadataAcross(coordinates, sources);
  if (metadata === undefined && versions.length === 0) {
    return { error: `package not found: ${groupId}:${artifactId}:${version}`, code: 1 };
  }

  // OSV scan of just this version; a network failure must not sink the card.
  let vulnerabilities: ShowAdvisory[] = [];
  try {
    const report = await auditPackages([coordinates], auditSource);
    vulnerabilities = (report.vulnerable[0]?.advisories ?? []).map(a => ({
      id: a.id,
      aliases: a.aliases,
      severity: a.severity,
      summary: a.summary,
      fixedVersions: a.fixedVersions,
      url: a.url,
    }));
    // OSV returns advisories in database order; show them worst severity first.
    vulnerabilities.sort(
      (a, b) => SEVERITY_ORDER.indexOf(a.severity) - SEVERITY_ORDER.indexOf(b.severity),
    );
  } catch {
    vulnerabilities = [];
  }

  const spdx = [
    ...new Set((metadata?.licenses ?? []).map(l => normalizeLicense(l.name, l.url))),
  ].filter(s => s !== undefined);

  return {
    groupId,
    artifactId,
    version,
    explicitVersion: parsed.version !== undefined,
    latestVersion: latest,
    versionCount: versions.length,
    newer: versions.includes(version) ? versions.length - 1 - versions.indexOf(version) : 0,
    // POM descriptions are often multi-line and indented; collapse to one line.
    description: metadata?.description?.replace(/\s+/g, " ").trim() || undefined,
    homepage: metadata?.homepage,
    scmUrl: metadata?.scmUrl,
    spdx,
    rawLicenses: (metadata?.licenses ?? []).map(l => l.name),
    dependencies: [...(metadata?.dependencies ?? [])].sort((a, b) =>
      `${a.groupId}:${a.artifactId}`.localeCompare(`${b.groupId}:${b.artifactId}`),
    ),
    project: projectContext(config, `${groupId}:${artifactId}`),
    vulnerabilities,
  };
}

/** The machine-readable view (one stable shape for --json). */
export function showToJson(data: ShowData): object {
  return {
    groupId: data.groupId,
    artifactId: data.artifactId,
    version: data.version,
    latestVersion: data.latestVersion ?? null,
    versionCount: data.versionCount,
    description: data.description ?? null,
    homepage: data.homepage ?? null,
    scmUrl: data.scmUrl ?? null,
    license: data.spdx.length > 0 ? data.spdx : data.rawLicenses,
    dependencies: data.dependencies.map(d => ({
      groupId: d.groupId,
      artifactId: d.artifactId,
      version: d.version,
      ...(d.scope ? { scope: d.scope } : {}),
      ...(d.optional ? { optional: true } : {}),
    })),
    project: {
      configurations: data.project.configurations,
      declared: data.project.declared ?? null,
      installed: data.project.installed ?? null,
    },
    vulnerabilities: data.vulnerabilities,
  };
}

type Paint = (format: StyleFormat, text: string) => string;

/** Render the colored detail card (paint is a no-op when color is disabled). */
export function renderShowCard(data: ShowData, paint: Paint): string {
  const lines: string[] = [];
  const row = (label: string, value: string): void => {
    lines.push(`  ${paint("dim", label.padEnd(LABEL_WIDTH))}${value}`);
  };

  // Header: coordinates + version, with a freshness hint from the version list.
  const hint =
    !data.explicitVersion || data.version === data.latestVersion
      ? paint("green", "latest")
      : data.newer > 0
        ? paint("yellow", `${data.newer} newer available`)
        : "";
  lines.push(
    `${paint(["bold", "cyan"], `${data.groupId}:${data.artifactId}`)} ${paint("bold", data.version)}` +
      (hint ? `  ${hint}` : ""),
  );
  if (data.description) lines.push(paint("dim", `  ${data.description}`));
  lines.push("");

  const licenseLabel =
    data.spdx.length > 0
      ? paint("cyan", data.spdx.join(", "))
      : data.rawLicenses.length > 0
        ? paint("yellow", `${data.rawLicenses.join(", ")} (no SPDX id)`)
        : paint("dim", "none declared");
  row("License", licenseLabel);
  if (data.homepage) row("Homepage", paint("blue", data.homepage));
  if (data.scmUrl) row("Repository", paint("blue", data.scmUrl));
  if (data.latestVersion !== undefined) {
    row(
      "Versions",
      `${data.latestVersion} ${paint("dim", "(latest)")}${paint("dim", `, ${data.versionCount} published`)}`,
    );
  }
  row("In project", formatProject(data.project, paint));

  lines.push("", `  ${paint("bold", `Dependencies (${data.dependencies.length})`)}`);
  if (data.dependencies.length === 0) {
    lines.push(`    ${paint("dim", "none")}`);
  } else {
    for (const d of data.dependencies) lines.push(`    ${formatDependency(d, paint)}`);
  }

  lines.push("", `  ${paint("bold", "Vulnerabilities")}`);
  if (data.vulnerabilities.length === 0) {
    lines.push(`    ${paint("green", "no known vulnerabilities")}`);
  } else {
    for (const a of data.vulnerabilities) {
      const cve = a.aliases.length > 0 ? ` (${a.aliases.join(", ")})` : "";
      const fixed =
        a.fixedVersions.length > 0
          ? paint("dim", `  [fixed in: ${a.fixedVersions.join(", ")}]`)
          : "";
      lines.push(
        `    ${paint(SEVERITY_STYLE[a.severity], a.severity.toUpperCase())}  ` +
          `${a.id}${cve} - ${a.summary}${fixed}`,
      );
      lines.push(`      ${paint("dim", a.url)}`);
    }
    lines.push(`    ${paint("dim", "run `cappu audit` to scan the whole dependency tree")}`);
  }

  return `${lines.join("\n")}\n`;
}

function formatProject(context: ProjectContext, paint: Paint): string {
  if (context.configurations.length === 0) return paint("dim", "not a direct dependency");
  const where = context.configurations.join(", ");
  const parts: string[] = [];
  if (context.declared !== undefined) parts.push(`declared ${context.declared}`);
  if (context.installed !== undefined) parts.push(`installed ${context.installed}`);
  return `${where}${parts.length > 0 ? paint("dim", ` (${parts.join(", ")})`) : ""}`;
}

function formatDependency(d: DependencyDeclaration, paint: Paint): string {
  const coord = `${d.groupId}:${d.artifactId}:${d.version}`;
  const tags: string[] = [];
  if (d.scope && d.scope !== "compile") tags.push(d.scope);
  if (d.optional) tags.push("optional");
  return `${coord}${tags.length > 0 ? paint("dim", `  ${tags.join(", ")}`) : ""}`;
}

export async function runShow(
  coord: string,
  config: CappuConfig,
  options: { json?: boolean } = {},
  sources: readonly PackageSource[] = configuredSources(config),
  auditSource: AuditSource = new OsvSource(cachedFetchJson()),
): Promise<never> {
  const data = await buildShowData(coord, config, sources, auditSource);
  if ("error" in data) {
    process.stderr.write(`cappu: ${data.error}\n`);
    process.exit(data.code);
  }

  if (options.json) {
    process.stdout.write(`${JSON.stringify(showToJson(data), null, 2)}\n`);
  } else {
    process.stdout.write(renderShowCard(data, painter(process.stdout)));
  }
  process.exit(data.vulnerabilities.length > 0 ? 1 : 0);
}
